authentication-via-social-networks/README.md

```diff
@@ -0,0 +1,1155 @@
+# Authentication via Social Networks
+
+Omniauth schema:
+![alt text](omniauth-schema.png "Omniauth schema")
+
+```ruby
+# Gemfile
+...
+gem "omniauth-facebook"
+gem "omniauth-google-oauth2"
+...
+```
+
+```ruby
+# app/controllers/omniauth_callbacks_controller.rb
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  SocialProfile::PROVIDERS.each do |provider|
+    define_method(provider.to_s) do
+      begin
+        current_user ? connect_social_profile : handle_sign_in
+      rescue AuthVerificationPolicy::OauthError => e
+        redirect_to root_path, flash: { notice: e.to_s }
+      end
+    end
+  end
+
+  private
+
+  def connect_social_profile
+    OauthConnectOrganizer.new(auth, current_user).call
+    redirect_to edit_user_registration_path
+  end
+
+  def auth
+    request.env["omniauth.auth"]
+  end
+
+  def handle_sign_in
+    user = auth_organizer.new(auth).user
+    sign_in_and_redirect user, event: :authentication
+  end
+
+  def auth_organizer
+    auth_verified? ? VerifiedAuthOrganizer : UnverifiedAuthOrganizer
+  end
+
+  def auth_verified?
+    AuthVerificationPolicy.new(auth).verified?
+  end
+
+  def after_sign_in_path_for(resource)
+    if resource.confirmed?
+      edit_user_registration_path
+    else
+      session[:auth_verified?] = auth_verified?
+      resource.reset_password(new_password, new_password)
+      finish_signup_path(resource)
+    end
+  end
+
+  def new_password
+    @new_password ||= Devise.friendly_token.first(8)
+  end
+end
+```
+
+```ruby
+# app/controllers/social_profiles_controller.rb
+class SocialProfilesController < ApplicationController
+  before_action :authenticate_user!
+
+  expose(:social_profiles) { current_user.social_profiles }
+  expose(:social_profile)
+
+  def destroy
+    if social_profile.destroy
+      flash[:notice] = t "flash.actions.destroy.notice", resource_name: resource_name
+    else
+      flash[:alert] = t "flash.actions.destroy.alert", resource_name: resource_name
+    end
+    redirect_to edit_user_registration_url
+  end
+
+  private
+
+  def resource_name
+    SocialProfile.model_name.human
+  end
+end
+```
+
+```ruby
+# app/controllers/users_controller.rb
+class UsersController < ApplicationController
+  before_action :authenticate_user!, only: :home
+
+  expose(:user, attributes: :user_params)
```

Comment:
@maximlarionov Main issue that we pass direct `id` into url. In this case user can manually change `id` param and update foreign users email and password attributes.